// termx/build/terminal.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var _a = require('chalk'), blue = _a.blue, bold = _a.bold, yellow = _a.yellow, red = _a.red, green = _a.green, magenta = _a.magenta, italic = _a.italic, cyan = _a.cyan;
__export(require("./boxes"));
//* Utils
exports.info = function (str) { return blue(str); };
exports.warning = function (str) { return bold(yellow(str)); };
exports.danger = function (str) { return bold(red(str)); };
exports.highlight = function (str) { return bold(green(str)); };
exports.field = function (str) { return italic(magenta(str)); };
exports.cold = function (str) { return bold(cyan(str)); };
//* Utils
exports.timestamp = function () { return cyan("[" + new Date().toUTCString() + "]"); };
//* Log
exports.log = function () {
    var data = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        data[_i] = arguments[_i];
    }
    console.log.apply(console, __spreadArrays([exports.timestamp()], data));
    return Date.now();
};
exports.warn = function (data) { return exports.log(exports.warning(data)); };
exports.line = function () { return console.log('\n'); };
